src/lib/kernel/Linux.cc

```diff
@@ -141,30 +142,32 @@ int64_t Linux::faccessat(int64_t dfd, const std::string& filename, int64_t mode,
   // special file)
   new_pathname = Linux::getSpecialFile(filename);
 
-  // Get correct dirfd
-  int64_t dirfd = Linux::getDirFd(dfd, filename);
-  if (dirfd == -1) return EBADF;
+  // Get host dirfd. May return -1 in case of no mapping, pass through to host
+  // faccessat to deal with this
+  int64_t hostDfd = Linux::getHostDirFD(dfd);
 
   // Pass call through to host
-  int64_t retval = ::faccessat(dirfd, new_pathname.c_str(), mode, flag);
+  int64_t retval = ::faccessat(hostDfd, new_pathname.c_str(), mode, flag);
 
   return retval;
 }
 
-int64_t Linux::close(int64_t fd) {
+int64_t Linux::close(int64_t vfd) {
   // Don't close STDOUT or STDERR otherwise no SimEng output is given
   // afterwards. This includes final results given at the end of execution
-  if (fd != STDERR_FILENO && fd != STDOUT_FILENO) {
-    assert(fd < processStates_[0].fileDescriptorTable.size());
-    int64_t hfd = processStates_[0].fileDescriptorTable[fd];
+  if (vfd != STDERR_FILENO && vfd != STDOUT_FILENO) {
+    assert(vfd >= 0 && static_cast<size_t>(vfd) <
+                           processStates_[0].fileDescriptorTable.size());
+    int64_t hfd = processStates_[0].fileDescriptorTable[vfd];
     if (hfd < 0) {
+      // Early return, can't deallocate vfd that isn't in fileDescriptorTable
       return EBADF;
     }
```

Comment:
Given similar logic was changed above (assert removed and size of fileDescriptorTable checked before access), why has it not been changed here?

Reply:
I left this as the previous implementation didn't use the getHostDirFD function and because the close syscall deals with "file descriptors" as opposed to "directory file descriptors". These seem to have slightly different functionality e.g. AT_FDCWD. Even though we store mappings in the same table I think there should be some separation. I think using the function here would give correct functionality however. Thoughts?